server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchMetadataFieldsPhase.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.fetch.subphase;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.common.document.DocumentField;
+import org.elasticsearch.search.fetch.FetchContext;
+import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.fetch.FetchSubPhaseProcessor;
+import org.elasticsearch.search.fetch.StoredFieldsContext;
+import org.elasticsearch.search.fetch.StoredFieldsSpec;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class FetchMetadataFieldsPhase implements FetchSubPhase {
+    @Override
+    public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
+        StoredFieldsContext storedFieldsContext = fetchContext.storedFieldsContext();
+        if (storedFieldsContext == null || storedFieldsContext.fetchFields() == false) {
```

Comment:
I see that MetadataFetcher below uses valueFetchers to retrieve values, and that is great. That means that this phase should no longer be exposed to how a certain field is stored, it just retrieves metadata fields via value fetchers. 

That should be straightforward for metadata fields which get returned by default, because they are not explicitly requested and instead of being retrieve from stored fields they go via value fetchers.

Metadata fields that are requested explicitly are requested in different ways: via `fields`, which goes through `FetchFieldsPhase`, and via `stored_fields` which currently goes through `StoredFieldsPhase`. It seems to me that you intend to move the latter and not the former to this new phase. In fact `FetchFieldsPhase` already goes through value fetchers, so requesting `fields:_ignored` would become equivalent to retrieving it via `stored_fields:_ignored` in how the field is retrieved, but it would be returned in a different section in the search response?

What is odd here is that for this phase to return metadata fields via value fetchers, they need to be requested via `stored_fields`. That tells me that it may make sense to unify this new phase with the existing FetchFieldsPhase, and add the notion of extracting meta fields requested via stored fields context. That would also require less code duplication, or am I missing something?